docs(metrics-operator): usage of SLI and SLO converters (#2013)
keptn/lifecycle-toolkit: metrics-operator/converter/sli_converter.md

# SLI Converter

## Description

SLI converter is a tool to convert `sli.yaml` files used in [KeptnV1](https://v1.keptn.sh/) into the new
`AnalysisValueTemplate` resources used in the new kubernetes-native [Keptn](https://lifecycle.keptn.sh/).
The converter is part of `metrics-operator` image.

## Usage

The converter will convert a single `sli.yaml` file into multiple `AnalysisValueTemplate` resources.

To run the converter, execute the following command:

> ```shell
> docker-run <METRICS_OPERATOR_IMAGE> manager --convert-sli=<PATH_TO_SLI> --sli-provider=<PROVIDER_NAME> --sli-namespace=<PROVIDER_NAMESPACE>

Reviewer: This is just personal preference, but listing the required variables to be set at the beginning and then use them as env vars in the command makes it a bit easier to see immediately what needs to be set - I usually tend overlook and forget at least one value to be set if the placeholders are spread over the command ;) 

```suggestion
```shell
METRICS_OPERATOR_IMAGE=<METRICS_OPERATOR_IMAGE>
PATH_TO_SLI=<PATH_TO_SLI>
PROVIDER_NAME=<PROVIDER_NAME>
PROVIDER_NAMESPACE=<PROVIDER_NAMESPACE>

docker-run $METRICS_OPERATOR_IMAGE manager --convert-sli=$PATH_TO_SLI --sli-provider=$PROVIDER_NAME --sli-namespace=$PROVIDER_NAMESPACE
```

```

Please be aware, you need to substitute the placeholders with the following information:

* **PATH_TO_SLI** - path to your `sli.yaml` file
* **PROVIDER_NAME** - name of `KeptnMetricsProvider` which will be used to fetch SLIs
* **PROVIDER_NAMESPACE** - namespace of `KeptnMetricsProvider` which will be used to fetch SLIs

> **Note**

> All the SLIs present in `sli.yaml` file will use the same provider defined by referenced

Reviewer: ```suggestion
All the SLIs present in `sli.yaml` file will use the same provider defined by the referenced
```

`KeptnMetricsProvider`.

## Example

The following content of `sli.yaml` file

```yaml
spec_version: "1.0"
indicators:
  throughput: "builtin:service.requestCount.total:merge(0):count?scope=tag(keptn_project:$PROJECT),tag(keptn_stage:$STAGE),tag(keptn_service:$SERVICE),tag(keptn_deployment:$DEPLOYMENT)"
  response_time_p95: "builtin:service.response.time:merge(0):percentile(95)?scope=tag(keptn_project:$PROJECT),tag(keptn_stage:$STAGE),tag(keptn_service:$SERVICE),tag(keptn_deployment:$DEPLOYMENT)"
```

> 

Reviewer: let's add the command that will create this output here to see where the namespace and provider name comes from in the resulting AVT, i.e.

```
docker-run $METRICS_OPERATOR_IMAGE manager --convert-sli=./sli.yaml --sli-provider=dynatrace --sli-namespace=keptn
```

# SLI Converter

SLI converter is a tool to convert the old `sli.yaml`
[file format](https://github.com/keptn/spec/blob/master/service_level_indicator.md) into the new
`AnalysisValueTemplate` [custom resource definition](../../docs/content/en/docs/crd-ref/metrics/v1alpha3/_index.md).
The converter is part of `metrics-operator` image.

## Usage

The converter will convert a single `sli.yaml` file into multiple `AnalysisValueTemplate` resources.

To run the converter, execute the following command:

```shell

> METRICS_OPERATOR_IMAGE=ghcr.io/keptn/lifecycle-operator:latest

Reviewer 2: we don't have any latest images. please set a real image version

Author: first we need to release then :)

Reviewer 2: metrics operator is already released now :P

Author: fixed

# SLI Converter

> ## Description

Reviewer 3: We can immediately with the description, no need for a headline :) 
```suggestion

```

> SLI converter is a tool to convert `sli.yaml` files used in [KeptnV1](https://v1.keptn.sh/) into the new
> `AnalysisValueTemplate` resources used in the new kubernetes-native [Keptn](https://lifecycle.keptn.sh/).

Reviewer 3: I won't use the v1 links since they will soon disappear 
```suggestion
SLI converter is a tool to convert the old `sli.yaml` [file format](https://github.com/keptn/spec/blob/master/service_level_indicator.md) into the new
`AnalysisValueTemplate` [custom resource definition](../../docs/content/en/docs/crd-ref/metrics/v1alpha3/_index.md).
```

PROVIDER_NAME=<PROVIDER_NAME>
PROVIDER_NAMESPACE=<PROVIDER_NAMESPACE>

docker-run $METRICS_OPERATOR_IMAGE manager --convert-sli=$PATH_TO_SLI --sli-provider=$PROVIDER_NAME --sli-namespace=$PROVIDER_NAMESPACE
```

Please be aware, you need to substitute the placeholders with the following information:

* **PATH_TO_SLI** - path to your `sli.yaml` file
* **PROVIDER_NAME** - name of `KeptnMetricsProvider` which will be used to fetch SLIs
* **PROVIDER_NAMESPACE** - namespace of `KeptnMetricsProvider` which will be used to fetch SLIs

> **Note**

All the SLIs present in `sli.yaml` file will use the same provider defined by the referenced
`KeptnMetricsProvider`.

## Example

The following content of `sli.yaml` file

```yaml
spec_version: "1.0"
indicators:
  throughput: "builtin:service.requestCount.total:merge(0):count?scope=tag(keptn_project:$PROJECT),tag(keptn_stage:$STAGE),tag(keptn_service:$SERVICE),tag(keptn_deployment:$DEPLOYMENT)"
  response_time_p95: "builtin:service.response.time:merge(0):percentile(95)?scope=tag(keptn_project:$PROJECT),tag(keptn_stage:$STAGE),tag(keptn_service:$SERVICE),tag(keptn_deployment:$DEPLOYMENT)"
```

with the following command

```shell

> docker-run $METRICS_OPERATOR_IMAGE manager --convert-sli=./sli.yaml --sli-provider=dynatrace --sli-namespace=keptn

Reviewer 3: maybe worth using `latest` as a tag here with the ghrc registry

# SLI Converter

SLI converter is a tool to convert the old `sli.yaml`
[file format](https://github.com/keptn/spec/blob/master/service_level_indicator.md) into the new
`AnalysisValueTemplate` [custom resource definition](../../docs/content/en/docs/crd-ref/metrics/v1alpha3/_index.md).
The converter is part of `metrics-operator` image.

## Usage

The converter will convert a single `sli.yaml` file into multiple `AnalysisValueTemplate` resources.

To run the converter, execute the following command:

> <!---x-release-please-start-version-->

Reviewer 3: I think you need to adapt release please conf to let it know that the action has to watch the files

Author: adapted

```shell
METRICS_OPERATOR_IMAGE=ghcr.io/keptn/metrics-operator:v0.8.2
PATH_TO_SLI=<PATH_TO_SLI>
KEPTN_PROVIDER_NAME=<KEPTN_PROVIDER_NAME>
KEPTN_PROVIDER_NAMESPACE=<KEPTN_PROVIDER_NAMESPACE>

> docker-run $METRICS_OPERATOR_IMAGE manager --convert-sli=$PATH_TO_SLI --keptn-provider-name=$KEPTN_PROVIDER_NAME --keptn-provider-namespace=$KEPTN_PROVIDER_NAMESPACE

Reviewer 3: ```suggestion
docker run $METRICS_OPERATOR_IMAGE manager --convert-sli=$PATH_TO_SLI --keptn-provider-name=$KEPTN_PROVIDER_NAME --keptn-provider-namespace=$KEPTN_PROVIDER_NAMESPACE
```

```
<!---x-release-please-end-->

Please be aware, you need to substitute the placeholders with the following information:

* **PATH_TO_SLI** - path to your `sli.yaml` file
* **KEPTN_PROVIDER_NAME** - name of `KeptnMetricsProvider` which will be used to fetch SLIs
* **KEPTN_PROVIDER_NAMESPACE** - namespace of `KeptnMetricsProvider` which will be used to fetch SLIs

> **Note**

All the SLIs present in `sli.yaml` file will use the same provider defined by the referenced
`KeptnMetricsProvider`.

## Example

The following content of `sli.yaml` file

```yaml
spec_version: "1.0"
indicators:
  throughput: "builtin:service.requestCount.total:merge(0):count?scope=tag(keptn_project:$PROJECT),tag(keptn_stage:$STAGE),tag(keptn_service:$SERVICE),tag(keptn_deployment:$DEPLOYMENT)"
  response_time_p95: "builtin:service.response.time:merge(0):percentile(95)?scope=tag(keptn_project:$PROJECT),tag(keptn_stage:$STAGE),tag(keptn_service:$SERVICE),tag(keptn_deployment:$DEPLOYMENT)"
```

with the following command

```shell

> docker-run $METRICS_OPERATOR_IMAGE manager --convert-sli=./sli.yaml --sli-provider=dynatrace --sli-namespace=keptn

Reviewer 3: ```suggestion
docker run $METRICS_OPERATOR_IMAGE manager --convert-sli=./sli.yaml --sli-provider=dynatrace --sli-namespace=keptn
```
